Fixes for JSONDecodeError and other misc issues.

alandtse opened the pull request:
Fixes #10 by explicitly checking for the device json.  Still may make sense to have every API call check that the log-in is still valid although the use of cookies may moot that.

This contains other fixes and improvements. 

1.  Lxml was not in the requirements.
2. Captcha had a typo in the configurator.
3. session.[mainArt][url] had a keyerror. Added a check for key existence before assignment.
4. Saving of session cookies.
5. More debug logs.
6. Configurator failure message and explanation on what to do with blank images.

EDIT: Added more features and updated the PR description.

alandtse commented:
As an aside, given #9 which I think you should still resolve, you have my permission to upstream my changes as Apache-2.0.

keatontaylor commented on media_player/alexa.py:
```diff
@@ -12,6 +12,9 @@
 import voluptuous as vol
 from bs4 import BeautifulSoup
+import pickle
+from threading import Lock
```
Eventually imports not native to home assistant need to be be added to the requirements and loaded during the setup_platform phase. Although this will all be solved with a custom pip module eventually.

alandtse replied:
My understanding is pickle is actually native to python3. Threading is gone.

keatontaylor commented on media_player/alexa.py:
```diff
@@ -250,7 +272,7 @@ def refresh(self, device):
                 self._media_artist = self._session['infoText']['subText1']
                 self._media_album_name = self._session['infoText']['subText2']
-                self._media_image_url = self._session['mainArt']['url']
+                self._media_image_url = self._session['mainArt']['url'] if 'url' in self._session['mainArt'] else None
```
I'd prefer to keep maximum line length to 80 characters. (I think this is the requirement for official components for HA.)

alandtse replied:
It is. I was being lazy to try to figure out if it resolved the Decode bug. A lot of the Amazon API access needs to be moved out to pypi anyway before this can be submitted so I wasn't running it though the HA linter.

keatontaylor commented on media_player/alexa.py:
```diff
@@ -496,8 +530,8 @@ def get_inputs(self, soup):
 
-    def login(self, cookies=None, captcha=None):
-
+    def test_loggedin(self, cookies=None):
```
This can eventually be converted to an auth decorator (probably when the AlexaAPI is moved to a pip module)

alandtse replied:
Learned something new today about auth decorators. It probably needs to be for all the API functions as I suggested in #10. Agree that probably makes more sense in the pip module.

keatontaylor commented on media_player/alexa.py:
```diff
@@ -12,6 +12,9 @@
 import voluptuous as vol
 from bs4 import BeautifulSoup
+import pickle
+from threading import Lock
```
This thread locking will not pass the requirements to submit to HA as an official component. We need to avoid locking/blocking the component thread at all costs.

alandtse replied:
@keatontaylor The lock was probably unnecessary but came from the other ha component is lifted the code from. That was a sensor so a lock was probably critical on the file access. Should be fine to remove it.

alandtse replied:
Removed.

keatontaylor commented:
Thank you for the PR! I finished classes Monday so will be back to working on this soon.

alandtse commented:
@keatontaylor I think I addressed your immediate concerns.  Let me know.

keatontaylor commented on media_player/alexa.py:
```diff
@@ -471,15 +493,28 @@ def device_state_attributes(self):
-        self.login()
+        self._cookiefile = hass.config.path("{}.pickle".format(ALEXA_DATA))
+        cookies = None
+        if exists(self._cookiefile):
```
hass.config.path already returns None if the file doesn't exist. So no need for the os.path exists() function here. 

You can simply do 

```
self._cookiefile = hass.config.path("{}.pickle".format(ALEXA_DATA))
cookies = None
if self._cookiefile:
```

alandtse commented:
@keatontaylor: Did some additional edits.

1. Version bumped to 0.8.0
2. Exists removed.
3. Found a Python 3.5 bug for JSONDecodeError and fixed it. 🤞
4. Fixed all HA linter errors since you want to submit this eventually.
